File: Scala/src/main/scala/uk/org/winder/maths/Factorial.scala
package uk.org.winder.maths

import scala.annotation.tailrec

object Factorial {
	val one = BigInt(1)
	val two = BigInt(2)

	private def validate(n:BigInt):Unit = {
		if (n < 0) throw new IllegalArgumentException("Parameter must be a non-negative integer.")
	}

	def iterativeWhile(n:BigInt):BigInt = {
		validate(n)
		var i = n
		var result = one
		while (i > 0) {
			result = result * i
			i = i - one
		}
		result
	}
	def iterativeWhile(n:Int):BigInt = iterativeWhile(BigInt(n))
	def iterativeWhile(n:Long):BigInt = iterativeWhile(BigInt(n))

	def iterativeFor(n:BigInt):BigInt = {
		validate(n)
		var result = one
		for (i <- one to n) { result *= i }
		result
	}
	def iterativeFor(n:Int):BigInt = iterativeFor(BigInt(n))
	def iterativeFor(n:Long):BigInt = iterativeFor(BigInt(n))

	def iterativeForeach(n:BigInt):BigInt = {
		validate(n)
		var result = one
		(two to n) foreach ((i:BigInt) => result *= i)
		result
	}
	def iterativeForeach(n:Int):BigInt = iterativeForeach(BigInt(n))
	def iterativeForeach(n:Long):BigInt = iterativeForeach(BigInt(n))

	def productive(n:BigInt):BigInt = {
		validate(n)
			(one to n).product
	}
	def productive(n:Int):BigInt = productive(BigInt(n))
	def productive(n:Long):BigInt = productive(BigInt(n))

	def reductive(n:BigInt):BigInt = {
		validate(n)
		if (n < two) one
		else two to n reduceLeft(_ * _)
	}
	def reductive(n:Int):BigInt = reductive(BigInt(n))
	def reductive(n:Long):BigInt = reductive(BigInt(n))

	def foldLeftive(n:BigInt):BigInt = {
		validate(n)
		(one /: (two to n))(_ * _)
	}
	def foldLeftive(n:Int):BigInt = foldLeftive(BigInt(n))
	def foldLeftive(n:Long):BigInt = foldLeftive(BigInt(n))

	def naïveRecursive(n:BigInt):BigInt = {
		validate(n)
		if (n < two) one
		else n * naïveRecursive(n - 1)
	}
	def naïveRecursive(n:Int):BigInt = naïveRecursive(BigInt(n))
	def naïveRecursive(n:Long):BigInt = naïveRecursive(BigInt(n))

	def tailRecursive(n:BigInt):BigInt = {
		validate(n)
		if (n < two) one
		else {
			@tailrec
			def iterate(x:BigInt = one, result:BigInt = one):BigInt =
				if (x > n) result
				else iterate(x + one, result * x)
			iterate()
		}
	}
	def tailRecursive(n:Int):BigInt = tailRecursive(BigInt(n))
	def tailRecursive(n:Long):BigInt = tailRecursive(BigInt(n))

}

File: Kotlin/src/main/kotlin/uk/org/winder/maths/factorial/factorial.kt
package uk.org.winder.maths.factorial

import java.math.BigInteger

val zero = BigInteger.ZERO
val one = BigInteger.ONE
val two = 2.bigint

fun validate(x:BigInteger):Unit {
  if (x < zero) { throw IllegalArgumentException() }
}

fun iterative(x:BigInteger):BigInteger {
  validate(x)
  var total = one
  (two rangeTo x).forEach{i -> total *= i}
  return total
}
fun iterative(x:Int):BigInteger = iterative(x.bigint)
fun iterative(x:Long):BigInteger = iterative(x.bigint)

fun reductive(x:BigInteger):BigInteger {
  validate(x)
  if (x < two) { return one }
  return (one rangeTo x).reduce{t, i -> t * i}
}
fun reductive(x:Int):BigInteger = reductive(x.bigint)
fun reductive(x:Long):BigInteger = reductive(x.bigint)

fun naïve_recursive(x:BigInteger):BigInteger {
  validate(x)
  if (x < two) { return one }
  return x * naïve_recursive(x - one)
}
fun naïve_recursive(x:Int):BigInteger = naïve_recursive(x.bigint)
fun naïve_recursive(x:Long):BigInteger = naïve_recursive(x.bigint)

// Use snake case to avoid conflict with Kotlin's tailRecursive function.
fun tail_recursive(x:BigInteger):BigInteger {
  validate(x)
  tailrec fun iterate(n:BigInteger, t:BigInteger=one):BigInteger {
    if (n < two) { return t }
    return iterate(n - one, n * t)
  }
  return iterate(x)
}
fun tail_recursive(x:Int):BigInteger = tail_recursive(x.bigint)
fun tail_recursive(x:Long):BigInteger = tail_recursive(x.bigint)

